lib/src/rules/close_sinks.dart

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.close_sinks;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:linter/src/linter.dart';
+import 'package:linter/src/utilities/dart_type_utilities.dart';
+
+typedef void _VisitVariableDeclaration(VariableDeclaration node);
+
+const _desc = r'Close instances of `dart.core.Sink`.';
+
+const _details = r'''
+
+**DO** Invoke `close` on instances of `dart.core.Sink` to avoid memory leaks and
+unexpected behaviors.
+
+**BAD:**
+```
+class A {
+  IOSink _sinkA;
+  void init(filename) {
+    _sinkA = new File(filename).openWrite(); // LINT
+  }
+}
+```
+
+**BAD:**
+```
+void someFunction() {
+  IOSink _sinkF; // LINT
+}
+```
+
+**GOOD:**
+```
+class B {
+  IOSink _sinkB;
+  void init(filename) {
+    _sinkB = new File(filename).openWrite(); // OK
+  }
+
+  void dispose(filename) {
+    _sinkB.close();
+  }
+}
+```
+
+**GOOD:**
+```
+void someFunctionOK() {
+  IOSink _sinkFOK; // OK
+  _sinkFOK.close();
+}
+```
+''';
+
+class CloseSinks extends LintRule {
+  CloseSinks() : super(
+      name: 'close_sinks',
+      description: _desc,
+      details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+}
+
+class Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  Visitor(this.rule);
+
+  @override
+  void visitVariableDeclarationStatement(VariableDeclarationStatement node) {
+    FunctionDeclaration function =
+        node.getAncestor((a) => a is FunctionDeclaration);
```

Comment:
The `function` can be `null` if the variable declaration is inside a MethodDeclaration. It might be better to use `FunctionBody`.